services/command-router-base/src/main/java/org/eclipse/hono/commandrouter/impl/kafka/KafkaBasedMappingAndDelegatingCommandHandler.java

```diff
@@ -83,12 +100,143 @@ public void mapAndDelegateIncomingCommandMessage(final KafkaConsumerRecord<Strin
+     * is invoked on command objects in the same order that the commands got added
+     * to the queue.
+     */
+    class CommandQueue {
+        private final Queue<KafkaBasedCommandContext> queue = new LinkedList<>();
+
+        /**
+         * Adds the given command to the queue.
+         *
+         * @param commandContext The context containing the command to add.
+         * @throws NullPointerException if commandContext is {@code null}.
+         */
+        public void add(final KafkaBasedCommandContext commandContext) {
+            Objects.requireNonNull(commandContext);
+            queue.add(commandContext);
+        }
+
+        /**
+         * Removes the given command from the queue.
+         * <p>
+         * To be used for commands for which processing resulted in an error
+         * and {@link #applySendCommandAction(KafkaBasedCommandContext, Supplier)}
+         * will not be invoked.
+         *
+         * @param commandContext The context containing the command to add.
+         * @throws NullPointerException if commandContext is {@code null}.
+         */
+        public void remove(final KafkaBasedCommandContext commandContext) {
+            Objects.requireNonNull(commandContext);
+            if (queue.remove(commandContext)) {
+                sendNextCommandInQueueIfPossible();
+            }
+        }
+
+        /**
+         * Either invokes the given sendAction directly if it is next-in-line or makes sure
+         * the action is invoked at a later point in time according to the order in which
+         * commands were originally received.
+         *
+         * @param commandContext The context of the command to apply the given sendAction for.
+         * @param sendActionSupplier The Supplier for the action to send the given command to the internal Command and
+         *                           Control API endpoint provided by protocol adapters.
+         * @return A future indicating the outcome of sending the command message.
+         * @throws NullPointerException if any of the parameters is {@code null}.
+         */
+        public Future<Void> applySendCommandAction(final KafkaBasedCommandContext commandContext,
+                final Supplier<Future<Void>> sendActionSupplier) {
+            Objects.requireNonNull(commandContext);
+            Objects.requireNonNull(sendActionSupplier);
+
+            final Promise<Void> resultPromise = Promise.promise();
+            final Pair<Supplier<Future<Void>>, Promise<Void>> sendActionSupplierAndResultPromise = Pair
+                    .of(sendActionSupplier, resultPromise);
+            if (commandContext.equals(queue.peek())) {
+                sendGivenCommandAndNextInQueueIfPossible(sendActionSupplierAndResultPromise);
+            } else if (!queue.contains(commandContext)) {
+                log.warn("command won't get sent - already removed from queue [{}]", commandContext.getCommand());
```

Comment:
How about also logging this warning in a span too?

Reply:
I currently don't see a scenario where this case can actually happen, so this is just meant as a safeguard here. I think logging is enough for that case. The command context span would also probably be already finished here.

Reply:
It's a very rare scenario. FMPOV such scenarios are very difficult to debug and hence a trace log would be very useful there too.  We can also ensure that the span is not yet closed by checking if the promise corresponding to the KEY_COMMAND_SEND_ACTION_SUPPLIER_AND_RESULT_PROMISE is not finished.

Reply:
I've made the method now more consistent overall by failing the result promise in this case, releasing the command context and also adding a trace log.